interpreter/functions/formatting.go

```diff
@@ -0,0 +1,164 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functions
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// FormatCallback is an interface that allows user-defined behavior
+// for formatting clause implementations, as well as argument retrieval.
+type FormatCallback interface {
+	String(ref.Val, string) (string, error)
+	Decimal(ref.Val, string) (string, error)
+	Fixed(*int) func(ref.Val, string) (string, error)
+	Scientific(*int) func(ref.Val, string) (string, error)
+	Binary(ref.Val, string) (string, error)
+	Hex(bool) func(ref.Val, string) (string, error)
+	Octal(ref.Val, string) (string, error)
+	Arg(int64) (ref.Val, error)
+	ArgSize() int64
+}
+
+type clauseImpl func(ref.Val, string) (string, error)
+
+// ParseFormatString takes a format string, FormatCallback implementation, and locale as input,
+// and returns a formatted string if formatting succeeds, or an error otherwise.
+func ParseFormatString(formatStr string, callback FormatCallback, locale string) (string, error) {
+	i := 0
+	argIndex := 0
+	var builtStr strings.Builder
+	for i < len(formatStr) {
+		if formatStr[i] == '%' {
+			if i+1 < len(formatStr) && formatStr[i+1] == '%' {
+				err := builtStr.WriteByte('%')
+				if err != nil {
+					return "", fmt.Errorf("error writing format string: %w", err)
+				}
+				i += 2
+				continue
+			} else {
+				newStrIndex, val, refErr := parseFormatArg(i, argIndex, formatStr, callback, locale)
+				if refErr != nil {
+					return "", refErr
+				}
+				_, err := builtStr.WriteString(val)
+				if err != nil {
+					return "", fmt.Errorf("error writing format string: %w", err)
+				}
+				i = newStrIndex
+				argIndex++
+			}
+		} else {
+			err := builtStr.WriteByte(formatStr[i])
+			if err != nil {
+				return "", fmt.Errorf("error writing format string: %w", err)
+			}
+			i++
+		}
+	}
+	return builtStr.String(), nil
+}
+
+// parseFormatArg takes the most recently-parsed string index, the next argument index to be used,
+// the complete formatting string,, a FormatCallback implementation, and the locale (or an
+// empty string if no locale was set). It returns the next index in the formatting string that
+// should be parsed, as well as a string containing the formatted argument, or an error if a problem
+// was encountered.
+func parseFormatArg(lastStrIndex, argIndex int, formatStr string, callback FormatCallback, locale string) (int, string, error) {
```

Comment:
I found this function difficult to understand at first. I believe it could be improved with some combination of:

- Godoc that says more about what the function does.
- A name that is slightly more descriptive (I found the use of the term "arg" confusing-- I think of the args as the values being interpolated into the string, but this seems to be more about the format clauses)
- Reduced coupling with `ParseFormatString`, particularly `lastStrIndex` and `argIndex`. Maybe:
  -  Eliminate `lastStrIndex` by using Go slices with something like `formatStr[i:]` to pass just the substring needed from `ParseFormatString` to `parseFormatArg`? Note that `parseFormatArg` will be returning a count of how many chars were consumed which `ParseFormatString` will then need to use as an offset.
  - Instead of passing in `argIndex`, figure out the right arg in `ParseFormatString` and then just pass in the arg?

Reply:
Makes sense, I've implemented each of those changes.